src/tree/updater_gpu_hist.cu

```diff
@@ -1375,6 +1403,58 @@ inline void DeviceShard<GradientSumT>::CreateHistIndices(
   entries_d.shrink_to_fit();
 }
 
+// An instance of this type is created which keeps track of total number of rows to process,
+// rows processed thus far, rows to process and the offset from the current sparse page batch
+// to begin processing on each device
+class DeviceHistogramBuilderState {
```

Comment:
I am wondering if there is another way to do this without having this class to manage the process. For example is it possible for each Shard to independently calculate its row state?

Reply:
i wanted to make this a part of the shard state to begin with, but decided against it. the state that is maintained by this type is transient in nature - meaning, this is needed only for creating the compressed bins during the tree initialization time. this isn't needed later during the tree construction/boosting time. further, the state maintained by this time repeatedly changes for each batch, and isn't even a true reflection of the entire dmatrix. thus, i decoupled this to be a separate type, as it isn't a part of the device shard, rather one used to construct its state.

Reply:
I'm ok with this.

Reply:
There's a lot of transient state in a shard. I think putting it inside a shard will make the code cleaner.

I'll leave it up to you, however.